pos-print/src/main/java/com/clouway/pos/print/adapter/db/BackgroundReceiptPrinterService.kt

```diff
@@ -0,0 +1,99 @@
+package com.clouway.pos.print.adapter.db
+
+import com.clouway.pos.print.core.*
+import com.clouway.pos.print.printer.Status
+import com.google.inject.Inject
+import com.mongodb.CursorType
+import com.mongodb.client.MongoCollection
+import org.bson.Document
+import org.slf4j.LoggerFactory
+import kotlin.concurrent.thread
+
+/**
+ * @author Tsvetozar Bonev (tsvetozar.bonev@clouway.com)
+ */
+class BackgroundReceiptPrinterService @Inject constructor(private var repository: ReceiptRepository,
+                                                          private var factory: PrinterFactory,
+                                                          private var notifier: PrintNotifier)
+  : BackgroundReceiptPrinter{
+
+  private val logger = LoggerFactory.getLogger(BackgroundReceiptPrinter::class.java)
+
+  override fun run() {
+    logger.info("Starting background printing service")
+    repository.getCappedCollection().drop()
+    thread {
+      while(true){
+        printReceipts(repository.getCappedCollection())
+      }
+    }
+  }
+
+
+  override fun printReceipts(collection: MongoCollection<Document>){
+    collection.find()
+      .cursorType(CursorType.TailableAwait)
+      .noCursorTimeout(true)
+      .iterator().use {
+        while (it.hasNext()) {
+          val receiptDoc = it.next()
+          val isFiscal = receiptDoc.getBoolean("isFiscal")
+          try{
+            val printer = factory.getPrinter(receiptDoc.getString("sourceIp"))
+
+            val receipt = adaptReceipt(receiptDoc)
+
+            val printResponse = if(isFiscal) printer.printFiscalReceipt(receipt)
+            else printer.printReceipt(receipt)
+
+            if(printResponse.warnings.contains(Status.FISCAL_RECEIPT_IS_OPEN)
+              || printResponse.warnings.contains(Status.NON_FISCAL_RECEIPT_IS_OPEN)){
+              logger.info("Receipt printing accepted")
+              repository.finishPrinting(receiptDoc["receiptId"].toString())
+            }else {
+              logger.info("Receipt printing rejected")
+              repository.rejectPrinting(receiptDoc["receiptId"].toString())
+            }
+
+            notifier.notifyPrinted(printResponse)
+
+          }catch (ex: DeviceNotFoundException){
+            logger.warn("Device was not found")
+            try {
+                repository.rejectPrinting(receiptDoc["receiptId"].toString())
+            }catch (ex: ReceiptNotInQueueException){
+              logger.warn("Receipt was not found in queue")
+            }
+          }catch (ex: ReceiptNotInQueueException){
+            logger.warn("Receipt was not found in queue")
+          }
+        }
+      }
+  }
+
+  @Suppress("UNCHECKED_CAST")
+  private fun adaptReceipt(receiptDoc: Document): Receipt {
```

Comment:
you should work with the documents only in the persistence adapter.
You can have PrinterQueue interfece 
```kotlin 
interface PrinterQueue {
    fun first():Receipt
    fun add(receipt: Receipt)
}
```
and you should work with it here in the logic

Reply:
Implemented a PrintQueue interface that provides methods to add a receipt, check for a next receipt, get the next receipt and clear itself.